src/algebra/quandle.lean

```diff
@@ -0,0 +1,561 @@
+def conj (G : Type*) := G
+
+@[simp] lemma conj_conj_inv {G : Type*} [group G] (g h : G) : g * (g⁻¹ * h * g) * g⁻¹ = h :=
+mul_inv_eq_of_eq_mul (mul_eq_of_eq_inv_mul (mul_assoc _ _ _))
+
+@[simp] lemma conj_inv_conj {G : Type*} [group G] (g h : G) : g⁻¹ * (g * h * g⁻¹) * g = h :=
+mul_eq_of_eq_mul_inv (inv_mul_eq_of_eq_mul (mul_assoc _ _ _))
+
+instance conj.quandle (G : Type*) [group G] : quandle (conj G) :=
+{ op' := (λ x, (@mul_aut.conj G _ x).to_equiv),
+  self_distrib' := λ x y z, by simp,
+  fix' := λ x, by simp }
+
+@[simp]
+lemma conj_op_eq_conj {G : Type*} [group G] (x y : conj G) :
+  x ◃ y = ((x : G) * (y : G) * (x : G)⁻¹ : G) := rfl
+
+lemma conj_swap {G : Type*} [group G] (x y : conj G) :
+  x ◃ y = y ↔ y ◃ x = x :=
+begin
+  dsimp, split,
+  repeat { intro h, conv_rhs { rw eq_mul_inv_of_mul_eq (eq_mul_inv_of_mul_eq h) }, simp, },
+end
+
+/--
+`conj` is functorial
+-/
+def conj.map {G : Type*} {H : Type*} [group G] [group H] (f : G →* H) : rack_hom (conj G) (conj H) :=
+{ to_fun := f,
+  map_op' := by simp }
+
+instance {G : Type*} {H : Type*} [group G] [group H] : has_lift (G →* H) (rack_hom (conj G) (conj H)) :=
+{ lift := conj.map }
+
+/--
+The dihedral quandle. This is the conjugation quandle of the dihedral group restrict to flips.
+
+Used for Fox n-colorings of knots.
+-/
+@[nolint has_inhabited_instance]
+def dihedral (n : ℕ) := zmod n
+
+/--
+The operation for the dihedral quandle.  It does not need to be an equivalence
+because it is an involution (see `dihedral_op.inv`).
+-/
+def dihedral_op (n : ℕ) (a : zmod n) : zmod n → zmod n :=
+λ b, 2 * a - b
+
+lemma dihedral_op.inv (n : ℕ) (a : zmod n) : function.involutive (dihedral_op n a) :=
+by { intro b, dsimp [dihedral_op], ring }
+
+instance (n : ℕ) : quandle (dihedral n) :=
+{ op' := λ a, function.involutive.to_equiv _ (dihedral_op.inv n a),
+  self_distrib' := λ x y z, begin
+    dsimp [function.involutive.to_equiv, dihedral_op], ring,
+  end,
+  fix' := λ x, begin
+    dsimp [op', function.involutive.to_equiv, dihedral_op], ring,
+  end }
+
+
+end quandle
+
+namespace rack
+
+/--
+This is the natural rack homomorphism to the conjugation quandle of the group `R ≃ R`
+that acts on the rack.
+-/
+def to_conj (R : Type*) [rack R] : rack_hom R (quandle.conj (R ≃ R)) :=
+{ to_fun := op',
+  map_op' := ad_conj }
+
+
+section envel_group
+
+/-!
+## Universal enveloping group of a rack
+
+Joyce, for quandles, called this AdConj.
+-/
+
+/--
+Free generators of the enveloping group.
+-/
+inductive pre_envel_group (R : Type u) : Type u
+| unit : pre_envel_group
+| incl (x : R) : pre_envel_group
+| mul (a b : pre_envel_group) : pre_envel_group
+| inv (a : pre_envel_group) : pre_envel_group
+
+instance pre_envel_group.inhabited (R : Type u) : inhabited (pre_envel_group R) :=
+⟨pre_envel_group.unit⟩
+
+open pre_envel_group
+
+/--
+Relations for the enveloping group.
+-/
+inductive pre_envel_group_rel' (R : Type u) [rack R] : pre_envel_group R → pre_envel_group R → Type u
```

Comment:
Why make this `Type` valued?

Reply:
It's because the `well_def` proof inducts on `pre_envel_group_rel'` to show that things related by `pre_envel_group_rel'` will map to the same element.  If it's a `Prop` then the induction isn't well-founded.

I think that's the only place it needed to be a `Type`.

Reply:
Since you do this, and then later make a specialised version in `Prop`, could you include the explanation of why you're doing this in comments? Otherwise it's going to be hard to decipher later.

Reply:
:+1: